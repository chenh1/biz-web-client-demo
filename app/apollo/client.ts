// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});