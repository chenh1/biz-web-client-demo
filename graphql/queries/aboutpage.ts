import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;