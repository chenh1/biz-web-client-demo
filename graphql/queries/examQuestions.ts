import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`