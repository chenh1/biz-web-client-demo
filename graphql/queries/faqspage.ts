import { gql, DocumentNode } from "@apollo/client";

const faqsPageQuery: DocumentNode = gql`
  query FaqsPage {
    faqsPage {
      data {
        attributes {
          title
          description
          faq(pagination: { page: 1, pageSize: 50 }) {
            question
            answer
            category {
              data {
                attributes {
                  displayName
                  value
                }
              }
            }
            answerRich
          }
        }
      }
    }
  }
`;

export default faqsPageQuery;