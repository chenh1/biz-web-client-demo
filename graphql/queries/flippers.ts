import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
