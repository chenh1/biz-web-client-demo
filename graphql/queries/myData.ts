import { gql, DocumentNode } from "@apollo/client";

export type ProgressData = {
  currentModule?: {
    lessonModule?: string;
  };
  completedModules?: Array<{
    lessonModule: string;
    examScore?: number;
  }>;
}

export type MyDataRes = {
  myData: {
    progressData: ProgressData
  } | {};
}

export const myDataQuery: DocumentNode = gql`
  query MyData {
    myData {
      progressData
    }
  }
`;
