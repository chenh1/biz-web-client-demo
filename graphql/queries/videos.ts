import { gql, DocumentNode } from "@apollo/client";

export type VideoType = {
  title: string;
  url: string;
  description: string;
  transcript: string;
  lessonModule: number | string;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
        value: string;
      }
    }[]
  }
}

export type Video = {
  attributes: VideoType;
}

export type VideoTypeResponse = {
  videos: {
    data: {
      attributes: VideoType
    }[]
  }
}

export const videosQuery: DocumentNode = gql`
  query Videos($firstModule: Float, $lastModule: Float) {
    videos(pagination: { page: 1, pageSize: 100 }, sort: "lessonModule:asc", filters: {lessonModule: {between: [ $firstModule, $lastModule ] }}) {
      data {
        attributes {
          title
          url
          description
          transcript
          lessonModule
          tags {
            data {
              attributes {
                displayName
                value
                color
              }
            }
          }
        }
      }
    }
  }
`