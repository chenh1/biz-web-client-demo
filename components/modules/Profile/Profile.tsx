'use client'

import type { CoreColors } from '@/components/constants/colors';
import type { SessionWithJwt } from '@/components/constants/types';
import { useSession } from 'next-auth/react';
import React, { useRef, useState } from 'react';
import { Overlay } from '../../containers/Overlay/Overlay';
import { Button } from '../../core/Button/Button';
import { Paragraph } from '../../core/Paragraph/Paragraph';
import { useQuery } from '@apollo/client';
import { myDataQuery } from '../../../graphql/queries/myData';
import { Divider } from '../../core/Divider/Divider';
import getScoreRank, { BRONZE, ranks } from '../../utils/getScoreRank';
import { Badge } from '../Deck/ExamView/Badge';
import { Heading } from '../../core/Heading/Heading';
import { Box } from '../../containers/Box/Box';
import Link from 'next/link';
import { videosQuery } from '../../../graphql/queries/videos';
import { RadioSet } from '../../fields/RadioSet/RadioSet';
import { archetypeIcons } from "../../constants/archetypes";

const AWARD_STATS = 'award_stats'
const MODULES_COMPLETED = 'modules_completed'

interface ProfileProps {
  setIsLoggingOut: (boolean) => void;
}

export const Profile: React.FC<ProfileProps> = ({ setIsLoggingOut }: ProfileProps) => {
  const session = useSession().data as SessionWithJwt | null;
  const [ showProfile, setShowProfile ] = useState(false)
  const [ achievementView, setAchievementView ] = useState(MODULES_COMPLETED)
  const awardStatsCheckboxRef = useRef(null)
  const completedModulesCheckboxRef = useRef(null)

  const { data: { myData: { progressData = {} } = {} } = {} } = useQuery(myDataQuery, { 
    context: { 
      headers: { 
        'Authorization': `Bearer ${session?.data?.jwt}` 
      },
    } ,
    skip: session?.status !== 'authenticated'
  })

  const { data: { videos: { data: videos = [] } = {} } = {} } = useQuery(videosQuery, { 
    variables: { firstModule: 0, lastModule: 999 } ,
    skip: session?.status !== 'authenticated'
  })
  const { completedModules = [] } = progressData || {};

  let awardStats = {}

  if (achievementView === AWARD_STATS) {
    completedModules.forEach(({ examScore }) => {
      const rank = getScoreRank(examScore)

      if (awardStats[rank]) {
        awardStats[rank] += 1
      } else {
        awardStats[rank] = 1
      }
    })
  }

  const iconUrl = archetypeIcons[progressData?.archetype]
  
  return (
    <div className="flex flex-col items-center justify-center w-full h-full">
      <div data-testid="toggleProfile" className="cursor-pointer relative w-12 h-12" onClick={() => setShowProfile(!showProfile)}>
        <div className="absolute w-full h-full rounded-full bg-gradient-to-br from-blue-400 to-blue-600">
          {iconUrl && <img title={`Greetings, ${progressData?.archetype}! You're logged in as ${session?.data?.email}`} src={iconUrl} alt="archetype icon" width={160} height={160}/>}
        </div>
      </div>

      <Overlay className="fixed top-24 right-4 bg-night-black-300 dark:bg-night-black-300 dark" shouldDisableBodyScroll={false} noDimmer noPopup visibility={showProfile} toggleVisibility={setShowProfile}>
        <Paragraph>Logged in as {session?.data?.email}</Paragraph>
        <Button onClick={()=> setIsLoggingOut(true)} label="Logout"/>
        <Divider/>

        <Box padding="0" className="grid-cols-2">
          <RadioSet
            onCheck={(value) => setAchievementView(value)}
            initialValue={achievementView}
            name="achievementView"
            data={[
              { label: 'Show award stats', value: AWARD_STATS },
              { label: 'Show completed modules', value: MODULES_COMPLETED }
            ]}
          />
        </Box>

        <div className="max-h-[50vh] overflow-y-auto">
          {achievementView === AWARD_STATS ?
            <Box padding="0" gap="sm" className="place-items-center">
              {ranks.map((rank, i) => (
                !!awardStats[rank] && (
                  <Box key={i} padding="0" className="place-items-center grid-cols-2">
                    <div className="scale-50 -m-6">
                      <Badge shouldLimitOverflow={false} rank={rank} />
                    </div>
                    <Heading type="h4" fontSizeClass="text-xl">x {awardStats[rank]}</Heading>
                  </Box>
                )
              ))}
            </Box>
            :
            <Box padding="0" gap="sm" className="justify-start">
              {completedModules.map(({ examScore, lessonModule }, i) => {
                const rank = getScoreRank(examScore)
                const scoreTextColor: `text-${CoreColors}` = rank === BRONZE ? 'text-max-white' : 'text-night-black-300'
                const scoreMessage = `${Math.round(examScore * 100)}`

                return (
                  <Box key={i} padding="0" className="justify-start items-center grid-cols-4">
                    <div className="scale-[.25] -m-12 col-span-1">
                      <Badge shouldLimitOverflow={false} rank={rank}>
                        <Heading type="h4" fontSizeClass="text-8xl" colorClass={scoreTextColor} darkColorClass={`dark:${scoreTextColor}`} className="z-50">
                          {scoreMessage}
                        </Heading>
                      </Badge>
                    </div>
                    <div className="col-span-3">
                      <Link href={`/module/${lessonModule.toString().replace('.', '-')}`}>
                        <Heading type="h4" fontSizeClass="text-xl">
                          {`${lessonModule}: ${videos?.length > 0 ? videos?.filter(({ attributes: { lessonModule: modNo } }) => lessonModule === modNo)[0]?.attributes?.title : ''}`}
                        </Heading>
                      </Link>
                    </div>
                  </Box>
                )
              })}
            </Box>
          }
        </div>
      </Overlay>
    </div>
  )
}