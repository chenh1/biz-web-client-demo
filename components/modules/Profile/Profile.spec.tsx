import { Profile } from "./Profile";
import { MockedProvider } from "@apollo/client/testing";
import { fireEvent, render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
import { myDataQuery } from "../../../graphql/queries/myData";
import { videosQuery } from "../../../graphql/queries/videos";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

const observe = jest.fn();
const unobserve = jest.fn();

// you can also pass the mock implementation
// to jest.fn as an argument
global.IntersectionObserver = jest.fn(function (callback, options) {
  return {
    observe,
    unobserve,
    root: null,
    rootMargin: '',
    thresholds: [],
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  };
});

jest.mock("next/link", () => ({ url, children }) => <a href={url}>children</a>);

const mocks = [
  {
    request: {
      query: myDataQuery,
    },
    result: {
      data: {
        myData: {
          progressData: {
            currentModule: {
              lessonModule: 1.01
            },
            completedModules: [
              { lessonModule: 1.01, examScore: 100 },
              { lessonModule: 1.02, examScore: 100 }
            ]
          }
        }
      }
    }
  },
  {
    request: {
      query: videosQuery,
    },
    result: {
      data: {
        videos: {
          data: [
            {
              attributes: {
                title: "Title 1",
                url: "https://www.youtube.com/watch?v=1",
                description: "Description 1",
                transcript: "Transcript 1",
                lessonModule: 1.01,
                tags: {
                  data: [
                    {
                      attributes: {
                        displayName: "Tag 1",
                        color: "red",
                        value: "tag1"
                      }
                    }
                  ]
                }
              }
            },
            {
              attributes: {
                title: "Title 2",
                url: "https://www.youtube.com/watch?v=2",
                description: "Description 2",
                transcript: "Transcript 2",
                lessonModule: 1.02,
                tags: {
                  data: [
                    {
                      attributes: {
                        displayName: "Tag 2",
                        color: "blue",
                        value: "tag2"
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
]

jest.mock('next-auth/react', () => ({
  useSession: jest.fn()
}))
import { useSession } from "next-auth/react";
const mockedUseSession = useSession as jest.Mock;


it('renders Profile module', async () => {
  mockedUseSession.mockImplementation(() => ({ status: "authenticated" }));

  const { getByTestId, container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Profile setIsLoggingOut={()=>{}}/>
    </MockedProvider>
  )

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));
  fireEvent.click(getByTestId('toggleProfile'));


  expect(container).toMatchSnapshot();
});
