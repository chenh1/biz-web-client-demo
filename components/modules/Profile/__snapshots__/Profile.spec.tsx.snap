// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Profile module 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center w-full h-full"
  >
    <div
      class="cursor-pointer relative w-12 h-12"
      data-testid="toggleProfile"
    >
      <div
        class="absolute w-full h-full rounded-full bg-gradient-to-br from-blue-400 to-blue-600"
      />
    </div>
    <div
      class="z-[10000] duration-300 ease-in-out opacity-100 fixed grid place-items-center left-0 top-0 w-screen h-screen "
      data-testid="overlayBg"
    >
      <div
        class="relative  freeze "
      >
        <div
          class=" grid rounded-md gap-4 shadow-lg p-8 bg-max-white dark:bg-night-black fixed top-24 right-4 bg-night-black-300 dark:bg-night-black-300 dark"
        >
          <p
            class=" text-night-black dark:text-max-white font-body font-body text-base undefined"
          >
            Logged in as 
          </p>
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5   "
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              Logout
            </span>
          </button>
          <hr
            class="w-full h-0 border-0 border-t-[1px] border-night-black-300 dark:border-max-white "
          />
          <div
            class=" grid  gap-4  p-2 grid-cols-2"
          >
            <div>
              <input
                class="  h-0 w-0"
                id="award_stats_0"
                name="achievementView"
                type="radio"
                value="award_stats"
              />
              <label
                class="  font-body relative flex my-[.6em] items-center text-night-black dark:text-max-white transition-colors duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)]"
                for="award_stats_0"
              >
                <span
                  class="pointer-events-none flex rounded-full justify-center items-center mr-[.9em] w-[1.1em] h-[1.1em] bg-transparent border-2 border-solid cursor-pointer transition-all duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)] before:content-[""] before:duration-300 before:origin-[0% 100%] before:w-[8px] before:h-[8px] before:absolute before:top-[0.49em] before:left-[0.3em] before:bg-night-black before:rounded-full before:scale-0 before:delay-300 border-night-black dark:border-max-white"
                />
                <span
                  class="pointer-events-none duration-500 opacity-1"
                >
                  Show award stats
                </span>
                <ins
                  class="pointer-events-none absolute block bottom-0 left-[2em] h-0 w-full overflow-hidden no-underline transition-[height] duration-300 ease-[cubic-bezier(.4,.0,.23,1)] "
                >
                  <i
                    class="pointer-events-none absolute bottom-0 not-italic text-hyper-red"
                  >
                    Show award stats
                  </i>
                </ins>
              </label>
            </div>
            <div>
              <input
                class="  h-0 w-0"
                id="modules_completed_1"
                name="achievementView"
                type="radio"
                value="modules_completed"
              />
              <label
                class="  font-body relative flex my-[.6em] items-center text-night-black dark:text-max-white transition-colors duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)]"
                for="modules_completed_1"
              >
                <span
                  class="pointer-events-none flex rounded-full justify-center items-center mr-[.9em] w-[1.1em] h-[1.1em] bg-transparent border-2 border-solid cursor-pointer transition-all duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)] before:content-[""] before:duration-300 before:origin-[0% 100%] before:w-[8px] before:h-[8px] before:absolute before:top-[0.49em] before:left-[0.3em] before:bg-night-black before:rounded-full before:scale-0 before:delay-300 border-[.55em] border-solid border-hyper-red animate-shrinkbounce before:scale-100"
                />
                <span
                  class="pointer-events-none duration-500 opacity-0"
                >
                  Show completed modules
                </span>
                <ins
                  class="pointer-events-none absolute block bottom-0 left-[2em] h-0 w-full overflow-hidden no-underline transition-[height] duration-300 ease-[cubic-bezier(.4,.0,.23,1)] h-full"
                >
                  <i
                    class="pointer-events-none absolute bottom-0 not-italic text-hyper-red"
                  >
                    Show completed modules
                  </i>
                </ins>
              </label>
            </div>
          </div>
          <div
            class="max-h-[50vh] overflow-y-auto"
          >
            <div
              class=" grid  gap-4  p-2 justify-start"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
